### ISSUE TITLE
Add retries to resources

### ISSUE THREAD

**Issue by gabcoyne** [PrefectHQ/terraform-provider-prefect#157]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave *+1* or *me too* comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### Terraform Version

Terraform version – 1.7.5
Prefect provider version – 0.1.2

### Affected Resource(s)

* prefect\_\*

### Terraform Configuration Files

```tf
terraform {
  required_providers {
    prefect = {
      source = "prefecthq/prefect"
    }
  }
}
  
provider "prefect" {
  account_id = var.prefect_account_id
  api_key = data.aws_secretsmanager_secret_version.api_token.secret_string
}
 
resource "prefect_workspace" "this" {
 name = var.name
  handle = lower(var.name)
}
```

### Expected Behavior

In the event of failure to create by 429 or 5XX errors, add retries to bring resiliency across the board.

### Actual Behavior

In this example a 503 was returned when attempting to create multiple workspaces at once

### Steps to Reproduce

1. `terraform apply`

### Resources

* [https://github.com/hashicorp/go-retryablehttp](https://github.com/hashicorp/go-retryablehttp)
* [https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/retries-and-customizable-timeouts](https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/retries-and-customizable-timeouts)
* [https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/retries-and-waiters.md](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/retries-and-waiters.md)

## Specifically-requested

* `prefect_work_pool` resource
* `prefect_workspace_role` datasource
* `prefect_worker_metadata` datasource
* `prefect_service_account` resource

**Comment by taylor-curran** [comment]
I have a plus one on this ticket from customer.

**Comment by mitchnielsen** [comment]
Received another report of this happening for `prefect_variable` resources.

**Comment by mitchnielsen** [comment]
We recently added retries for `prefect_workspace` resources in https://github.com/PrefectHQ/terraform-provider-prefect/pull/242. We can look into doing this for variables as well.

**Comment by mitchnielsen** [comment]
Workspaces and variables seem to be the two resources where this popped up the most. For now, rather than adding retries to _all_ resources, let's stick with what we've added so far. If anyone notices similar problems on other resources, please leave a comment here and we'll reopen to investigate.

**Comment by mch-sb** [comment]
Hey!
We are seeing the same issue with a couple of other resources:

## resource `prefect_work_pool`

```
│   on modules/prefect_work_pool_ecs/prefect_work_pool.tf line 99, in resource "prefect_work_pool" "project_work_pool":
│   99: resource "prefect_work_pool" "project_work_pool" {
│ 
│ Could not get Work Pool, unexpected error: status code 503 Service
│ Unavailable, error=
╵

Error: Process completed with exit code 1.
```

## data `prefect_workspace_role`

```
│   on modules/prefect/main.tf line 12, in data "prefect_workspace_role" "developer":
│   12: data "prefect_workspace_role" "developer" {
│ 
│ Could not list Workspace Role, unexpected error: status code 503 Service
│ Unavailable, error=upstream connect error or disconnect/reset before
│ headers. reset reason: connection termination
╵
Error: Process completed with exit code 1.

```

These fail every once in a while.

It would be quite nice if all resource/datasource had retries, as it seems to be possible to get this issue for most resources/datasources.

**Comment by mitchnielsen** [comment]
Thanks for the feedback! I'll reopen this one.

**Comment by erlendhbarstad** [comment]
I have another one for:
```
│    5: resource "prefect_service_account" "bot" {
│ 
```

Appreciate!

**Comment by mch-sb** [comment]
We have a new one:

```
│    8: resource "prefect_workspace_access" "bot_developer" {
│ 
│ Could not read Workspace Access, unexpected error: status code 503 Service
│ Unavailable, error=upstream connect error or disconnect/reset before
│ headers. reset reason: connection termination
```

**Comment by mitchnielsen** [comment]
We've released [v2.4.0](https://github.com/PrefectHQ/terraform-provider-prefect/releases/tag/v2.4.0), which includes a change that will now retry HTTP requests under the hood in specific scenarios. This applies to all resources and data sources.

We have pretty fine-grained control of when to retry and when not to, so please reach out and open an issue if there are any specific cases we can address.

Thanks everyone for the reports and feedback 🤝
